### ISSUE TITLE
XML Validation error after upgrade to v0.7.0

### ISSUE THREAD

**Issue by NickTracker** [ChuckJonas/vscode-apex-pmd#164]
**Describe the bug**
After upgrading to v0.7.0 of the Apex PMD extension the Apex PMD extension began throwing XML Validation errors. Downgrading to 0.6.2 resolves the issue.

` ^^^^^ net.sourceforge.pmd.lang.rule.XPathRule`

**Apex PMD Output**
```
###################################
Analyzing /Users/nzozaya/Salesforce/strk-core/src/main/default/classes/tests/builders/StSObjectBuilder.cls
env: {"CLASSPATH":"/Users/nzozaya/Salesforce/strk-core/*"}
PMD Command: "/Users/nzozaya@sitetracker.com/.vscode/extensions/chuckjonas.apex-pmd-0.7.0/bin/pmd/bin/pmd" check --no-progress -f csv --cache "/Users/nzozaya/Salesforce/strk-core/.pmdCache" -d "/Users/nzozaya/Salesforce/strk-core/src/main/default/classes/tests/builders/StSObjectBuilder.cls" -R "/Users/nzozaya/Salesforce/strk-core/strk-apex-standards.xml"
stderr:[ERROR] Error at /Users/nzozaya/Salesforce/strk-core/strk-apex-standards.xml:17:9
 15|         language="apex"
 16|         message="Use more descriptive variable names rather than short forms."
 17|         class="net.sourceforge.pmd.lang.rule.XPathRule">
             ^^^^^ net.sourceforge.pmd.lang.rule.XPathRule

 18|         <description>
 19|             Check for variable name under 4 characters or acronyms following one or less initial: net.sourceforge.pmd.lang.rule.XPathRule
stderr:[ERROR] Error at /Users/nzozaya/Salesforce/strk-core/strk-apex-standards.xml:42:9
 40|         language="apex"
 41|         message="Consider refactoring this 'if' statement for early return to simplify method structure."
 42|         class="net.sourceforge.pmd.lang.rule.XPathRule">
             ^^^^^ net.sourceforge.pmd.lang.rule.XPathRule

 43|         <description>
 44|             This rule detects 'if' statements within methods that are followed immediately by a: net.sourceforge.pmd.lang.rule.XPathRule
stderr:[ERROR] Error at /Users/nzozaya/Salesforce/strk-core/strk-apex-standards.xml:67:9
 65|         message="Methods annotated with @RemoteAction or @AuraEnabled must have a Model or StResult return
 66|     type"
 67|         class="net.sourceforge.pmd.lang.rule.XPathRule">
             ^^^^^ net.sourceforge.pmd.lang.rule.XPathRule

 68|         <description>
 69|             This rule enforces a naming convention on the return types of methods annotated with: net.sourceforge.pmd.lang.rule.XPathRule
stderr:[ERROR] Error at /Users/nzozaya/Salesforce/strk-core/strk-apex-standards.xml:94:9
 92|         language="apex"
 93|         message="Within methods annotated with @AuraEnabled or @RemoteAction, prefer throwing exceptions in response to unexpected inputs."
 94|         class="net.sourceforge.pmd.lang.rule.XPathRule">
             ^^^^^ net.sourceforge.pmd.lang.rule.XPathRule

 95|         <description>
 96|             This rule flags 'if' blocks within methods annotated with @AuraEnabled or @RemoteAction: net.sourceforge.pmd.lang.rule.XPathRule
stderr:[ERROR] Error at /Users/nzozaya/Salesforce/strk-core/strk-apex-standards.xml:121:9
 119|         language="apex"
 120|         message="Direct DML operations should be avoided in favor of Database methods using AccessLevel.USER_MODE or StDmlUtils to ensure security checks are performed."
 121|         class="net.sourceforge.pmd.lang.rule.XPathRule">
              ^^^^^ net.sourceforge.pmd.lang.rule.XPathRule

 122|         <description>
 123|             This rule highlights direct usage of DML operations like insert, update, delete, and: net.sourceforge.pmd.lang.rule.XPathRule
stderr:[ERROR] Error at /Users/nzozaya/Salesforce/strk-core/strk-apex-standards.xml:153:9
 151|         language="apex"
 152|         message="Ensure that test methods annotated with @isTest invoke Test.startTest() to properly manage test context and governor limits"
 153|         class="net.sourceforge.pmd.lang.rule.XPathRule">
              ^^^^^ net.sourceforge.pmd.lang.rule.XPathRule

 154|         <description>
 155|             This rule ensures that Apex test methods annotated with @isTest include an invocation of: net.sourceforge.pmd.lang.rule.XPathRule
stderr:[ERROR] Error at /Users/nzozaya/Salesforce/strk-core/strk-apex-standards.xml:181:9
 179|         language="apex"
 180|         message="Test methods annotated with @isTest should include a System.runAs block to ensure FLS and CRUD is respected within specific user contexts."
 181|         class="net.sourceforge.pmd.lang.rule.XPathRule">
              ^^^^^ net.sourceforge.pmd.lang.rule.XPathRule

 182|         <description>
 183|             This rule checks that Apex test methods annotated with @isTest properly simulate user: net.sourceforge.pmd.lang.rule.XPathRule
stderr:[ERROR] Error at /Users/nzozaya/Salesforce/strk-core/strk-apex-standards.xml:207:9
 205|         language="apex"
 206|         message="Test methods should be preceded by a formal comment explaining their purpose and context."
 207|         class="net.sourceforge.pmd.lang.rule.XPathRule">
              ^^^^^ net.sourceforge.pmd.lang.rule.XPathRule

 208|         <description>
 209|             Ensures that each test method annotated with @isTest has a preceding formal comment.: net.sourceforge.pmd.lang.rule.XPathRule
stderr:[ERROR] Error at /Users/nzozaya/Salesforce/strk-core/strk-apex-standards.xml:234:9
 232|         language="apex"
 233|         message="Avoid returning 'this' from methods to prevent chaining."
 234|         class="net.sourceforge.pmd.lang.rule.XPathRule">
              ^^^^^ net.sourceforge.pmd.lang.rule.XPathRule

 235|         <description>
 236|             This rule detects methods that return the current class instance using 'return this;',: net.sourceforge.pmd.lang.rule.XPathRule
stderr:[ERROR] Cannot load ruleset /Users/nzozaya/Salesforce/strk-core/strk-apex-standards.xml: 9 XML validation errors occurred
Failed Exit Code: 1
###################################
Analyzing /Users/nzozaya/Salesforce/strk-core/src/main/default/classes/tests/builders/StSObjectBuilder.cls
env: {"CLASSPATH":"/Users/nzozaya/Salesforce/strk-core/*"}
PMD Command: "/Users/nzozaya@sitetracker.com/.vscode/extensions/chuckjonas.apex-pmd-0.7.0/bin/pmd/bin/pmd" check --no-progress -f csv --cache "/Users/nzozaya/Salesforce/strk-core/.pmdCache" -d "/Users/nzozaya/Salesforce/strk-core/src/main/default/classes/tests/builders/StSObjectBuilder.cls" -R "/Users/nzozaya/Salesforce/strk-core/strk-apex-standards.xml"
stderr:[ERROR] Error at /Users/nzozaya/Salesforce/strk-core/strk-apex-standards.xml:17:9
 15|         language="apex"
 16|         message="Use more descriptive variable names rather than short forms."
 17|         class="net.sourceforge.pmd.lang.rule.XPathRule">
             ^^^^^ net.sourceforge.pmd.lang.rule.XPathRule

 18|         <description>
 19|             Check for variable name under 4 characters or acronyms following one or less initial: net.sourceforge.pmd.lang.rule.XPathRule
stderr:[ERROR] Error at /Users/nzozaya/Salesforce/strk-core/strk-apex-standards.xml:42:9
 40|         language="apex"
 41|         message="Consider refactoring this 'if' statement for early return to simplify method structure."
 42|         class="net.sourceforge.pmd.lang.rule.XPathRule">
             ^^^^^ net.sourceforge.pmd.lang.rule.XPathRule

 43|         <description>
 44|             This rule detects 'if' statements within methods that are followed immediately by a: net.sourceforge.pmd.lang.rule.XPathRule
stderr:[ERROR] Error at /Users/nzozaya/Salesforce/strk-core/strk-apex-standards.xml:67:9
 65|         message="Methods annotated with @RemoteAction or @AuraEnabled must have a Model or StResult return
 66|     type"
 67|         class="net.sourceforge.pmd.lang.rule.XPathRule">
             ^^^^^ net.sourceforge.pmd.lang.rule.XPathRule

 68|         <description>
 69|             This rule enforces a naming convention on the return types of methods annotated with: net.sourceforge.pmd.lang.rule.XPathRule
stderr:[ERROR] Error at /Users/nzozaya/Salesforce/strk-core/strk-apex-standards.xml:94:9
 92|         language="apex"
 93|         message="Within methods annotated with @AuraEnabled or @RemoteAction, prefer throwing exceptions in response to unexpected inputs."
 94|         class="net.sourceforge.pmd.lang.rule.XPathRule">
             ^^^^^ net.sourceforge.pmd.lang.rule.XPathRule

 95|         <description>
 96|             This rule flags 'if' blocks within methods annotated with @AuraEnabled or @RemoteAction: net.sourceforge.pmd.lang.rule.XPathRule
stderr:[ERROR] Error at /Users/nzozaya/Salesforce/strk-core/strk-apex-standards.xml:121:9
 119|         language="apex"
 120|         message="Direct DML operations should be avoided in favor of Database methods using AccessLevel.USER_MODE or StDmlUtils to ensure security checks are performed."
 121|         class="net.sourceforge.pmd.lang.rule.XPathRule">
              ^^^^^ net.sourceforge.pmd.lang.rule.XPathRule

 122|         <description>
 123|             This rule highlights direct usage of DML operations like insert, update, delete, and: net.sourceforge.pmd.lang.rule.XPathRule
stderr:[ERROR] Error at /Users/nzozaya/Salesforce/strk-core/strk-apex-standards.xml:153:9
 151|         language="apex"
 152|         message="Ensure that test methods annotated with @isTest invoke Test.startTest() to properly manage test context and governor limits"
 153|         class="net.sourceforge.pmd.lang.rule.XPathRule">
              ^^^^^ net.sourceforge.pmd.lang.rule.XPathRule

 154|         <description>
 155|             This rule ensures that Apex test methods annotated with @isTest include an invocation of: net.sourceforge.pmd.lang.rule.XPathRule
stderr:[ERROR] Error at /Users/nzozaya/Salesforce/strk-core/strk-apex-standards.xml:181:9
 179|         language="apex"
 180|         message="Test methods annotated with @isTest should include a System.runAs block to ensure FLS and CRUD is respected within specific user contexts."
 181|         class="net.sourceforge.pmd.lang.rule.XPathRule">
              ^^^^^ net.sourceforge.pmd.lang.rule.XPathRule

 182|         <description>
 183|             This rule checks that Apex test methods annotated with @isTest properly simulate user: net.sourceforge.pmd.lang.rule.XPathRule
stderr:[ERROR] Error at /Users/nzozaya/Salesforce/strk-core/strk-apex-standards.xml:207:9
 205|         language="apex"
 206|         message="Test methods should be preceded by a formal comment explaining their purpose and context."
 207|         class="net.sourceforge.pmd.lang.rule.XPathRule">
              ^^^^^ net.sourceforge.pmd.lang.rule.XPathRule

 208|         <description>
 209|             Ensures that each test method annotated with @isTest has a preceding formal comment.: net.sourceforge.pmd.lang.rule.XPathRule
stderr:[ERROR] Error at /Users/nzozaya/Salesforce/strk-core/strk-apex-standards.xml:234:9
 232|         language="apex"
 233|         message="Avoid returning 'this' from methods to prevent chaining."
 234|         class="net.sourceforge.pmd.lang.rule.XPathRule">
              ^^^^^ net.sourceforge.pmd.lang.rule.XPathRule

 235|         <description>
 236|             This rule detects methods that return the current class instance using 'return this;',: net.sourceforge.pmd.lang.rule.XPathRule
stderr:[ERROR] Cannot load ruleset /Users/nzozaya/Salesforce/strk-core/strk-apex-standards.xml: 9 XML validation errors occurred
Failed Exit Code: 1
```
**Apex PMD Settings**
```
<?xml version="1.0"?>
<!-- 
    Rules for enforcing our Dev Standards
https://sitetraker.atlassian.net/wiki/spaces/MMEP/pages/2874114053/Development+and+Test+Standards
 -->
<ruleset name="NSYNC Apex Standards"
    xmlns="http://pmd.sourceforge.net/ruleset/2.0.0"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xsi:schemaLocation="http://pmd.sourceforge.net/ruleset/2.0.0 https://pmd.sourceforge.io/ruleset_2_0_0.xsd">
    <description>
        These rules enforce the NSYNC team's documented best practices at compile time.
    </description>
    <rule
        name="DescriptiveVariableNameRule"
        language="apex"
        message="Use more descriptive variable names rather than short forms."
        class="net.sourceforge.pmd.lang.rule.XPathRule">
        <description>
            Check for variable name under 4 characters or acronyms following one or less initial
            small-case letters.
        </description>
        <priority>2</priority>
        <properties>
            <property name="xpath">
                <value>
                    <![CDATA[
                    //VariableDeclaration
                    [
                        string-length(@Image) < 4
                        or
                        matches(@Image, '^[a-z]{0,1}[A-Z]')
                    ]
                    ]]>
                </value>
            </property>
        </properties>
    </rule>
    <rule
        name="DetectEarlyReturnOpportunities"
        language="apex"
        message="Consider refactoring this 'if' statement for early return to simplify method structure."
        class="net.sourceforge.pmd.lang.rule.XPathRule">
        <description>
            This rule detects 'if' statements within methods that are followed immediately by a
            'ReturnStatement' or are the last statements in the method block, indicating a potential
            opportunity for an early return pattern, improving readability and reducing nesting.
        </description>
        <priority>2</priority>
        <properties>
            <property name="xpath">
                <value><![CDATA[
            //Method/BlockStatement/*[self::IfElseBlockStatement]
            [
                (following-sibling::*[1][self::ReturnStatement]) or
                (not(following-sibling::*) and parent::BlockStatement[last()])
            ]
            ]]></value>
            </property>
        </properties>
    </rule>

    <rule
        name="EnsureProperReturnTypeForApiMethods"
        language="apex"
        message="Methods annotated with @RemoteAction or @AuraEnabled must have a Model or StResult return
    type"
        class="net.sourceforge.pmd.lang.rule.XPathRule">
        <description>
            This rule enforces a naming convention on the return types of methods annotated with
            @RemoteAction or @AuraEnabled. The return type should either end with 'Model' or be
            exactly 'StResult'
        </description>
        <priority>2</priority>
        <properties>
            <property name="xpath">
                <value>
                <![CDATA[
                //Method[
                    (
                        ModifierNode/Annotation[@Name='RemoteAction' or @Name='AuraEnabled']
                    ) and (
                        not(@ReturnType = 'StResult' or contains(@ReturnType, 'Model'))
                    )
                ]
                ]]>
            </value>
            </property>
        </properties>
    </rule>
    <rule
        name="DetectEmptyListReturnsInAuraAndRemoteActionIfBlocks"
        language="apex"
        message="Within methods annotated with @AuraEnabled or @RemoteAction, prefer throwing exceptions in response to unexpected inputs."
        class="net.sourceforge.pmd.lang.rule.XPathRule">
        <description>
            This rule flags 'if' blocks within methods annotated with @AuraEnabled or @RemoteAction
            that return blank lists or null values which may be misleading to invokers of the API
        </description>
        <priority>3</priority>
        <properties>
            <property name="xpath">
                <value>
                <![CDATA[
                //Method[
                    ModifierNode/Annotation[@Name='AuraEnabled' or @Name='RemoteAction']
                ]
                /BlockStatement//IfElseBlockStatement/IfBlockStatement[
                    .//ReturnStatement[
                        NewListInitExpression or LiteralExpression[@LiteralType='NULL']
                    ]
                ]
                ]]>
            </value>
            </property>
        </properties>
    </rule>
    <rule
        name="EnforceSecureDML"
        language="apex"
        message="Direct DML operations should be avoided in favor of Database methods using AccessLevel.USER_MODE or StDmlUtils to ensure security checks are performed."
        class="net.sourceforge.pmd.lang.rule.XPathRule">
        <description>
            This rule highlights direct usage of DML operations like insert, update, delete, and
            upsert that are not protected by security-enhancing utility methods like Database
            actions or StDmlUtils. Highlighting specific DML statements helps developers quickly
            identify and rectify potentially unsafe code.
        </description>
        <priority>3</priority>
        <properties>
            <property name="xpath">
                <value><![CDATA[
                //Method[
                    not(
                        .//MethodCallExpression[
                            PrimaryExpression/Name[@Name='Database' or contains(@Name, 'StDmlUtils')]
                        ]
                    )
                ]/(
                    .//DmlInsertStatement |
                    .//DmlUpdateStatement |
                    .//DmlDeleteStatement |
                    .//DmlUpsertStatement
                )
            ]]></value>
            </property>
        </properties>
    </rule>
    <!-- Apex test rules -->
    <rule
        name="RequireTestStartTestInTestMethods"
        language="apex"
        message="Ensure that test methods annotated with @isTest invoke Test.startTest() to properly manage test context and governor limits"
        class="net.sourceforge.pmd.lang.rule.XPathRule">
        <description>
            This rule ensures that Apex test methods annotated with @isTest include an invocation of
            Test.startTest(). This practice is crucial for managing governor limits effectively by
            marking the test's main execution context.
        </description>
        <priority>3</priority>
        <properties>
            <property name="xpath">
                <value>
                <![CDATA[
                //Method[
                    ModifierNode/Annotation[@Name='IsTest']
                    and not(
                        .//MethodCallExpression[
                            @MethodName='startTest'
                        ]
                    )
                ]
                ]]>
            </value>
            </property>
        </properties>
    </rule>
    <rule
        name="EnsureRunAsInTestMethods"
        language="apex"
        message="Test methods annotated with @isTest should include a System.runAs block to ensure FLS and CRUD is respected within specific user contexts."
        class="net.sourceforge.pmd.lang.rule.XPathRule">
        <description>
            This rule checks that Apex test methods annotated with @isTest properly simulate user
            contexts by including System.runAs method calls. This is essential for testing features
            that depend on user permissions.
        </description>
        <priority>3</priority>
        <properties>
            <property name="xpath">
                <value>
                <![CDATA[
                //Method[
                    ModifierNode/Annotation[@Name='IsTest'] and
                    not(
                        .//RunAsBlockStatement
                    )
                ]
                ]]>
            </value>
            </property>
        </properties>
    </rule>
    <rule
        name="RequireDocumentationForTestMethods"
        language="apex"
        message="Test methods should be preceded by a formal comment explaining their purpose and context."
        class="net.sourceforge.pmd.lang.rule.XPathRule">
        <description>
            Ensures that each test method annotated with @isTest has a preceding formal comment.
            This comment is essential for explaining the test's purpose, providing context, or
            linking to relevant issue trackers.
        </description>
        <priority>3</priority>
        <properties>
            <property name="xpath">
                <value><![CDATA[
                    //Method[
                        ModifierNode/Annotation[@Name='IsTest'] and
                        not(
                            FormalComment[
                                contains(@Image, 'https') and 
                                contains(@Image, '|')
                            ]
                        )
                    ]             
            ]]></value>
            </property>
        </properties>
    </rule>
    <rule
        name="AvoidReturnThis"
        language="apex"
        message="Avoid returning 'this' from methods to prevent chaining."
        class="net.sourceforge.pmd.lang.rule.XPathRule">
        <description>
            This rule detects methods that return the current class instance using 'return this;',
            which can promote method chaining and potentially lead to less readable code.
        </description>
        <priority>3</priority>
        <properties>
            <property name="xpath">
                <value><![CDATA[
                //Method/BlockStatement//ReturnStatement
                    [ThisVariableExpression]
            ]]></value>
            </property>
        </properties>
    </rule>
</ruleset>
```

- pmdBinPath: n/a
- rulesets: `./strk-apex-standards.xml`
- additionalClassPaths: n/a

**Apex File**

**Screenshots**
<img width="1390" alt="Screenshot 2024-04-28 at 18 23 48" src="https://github.com/ChuckJonas/vscode-apex-pmd/assets/145612031/8c1faa92-9e5a-4624-95e6-d73f7e233fd8">

**Versions (please complete the following information):**
 - OS: `Darwin arm64 22.1.0`
 - vscode: `1.88.1`

**Comment by adangel** [comment]
The update to 0.7.0 of the plugin contained the upgrade to the final PMD 7.0.0 version. The switch from pre-7.0.0 to 7.0.0 is a major version change with known non backwards compatible changes.

Please refer to the [Migration Guide for PMD 7](https://docs.pmd-code.org/latest/pmd_userdocs_migrating_to_pmd7.html), especially the section [XPath rules](https://docs.pmd-code.org/latest/pmd_userdocs_migrating_to_pmd7.html#xpath-rules).

In short:
- replace `net.sourceforge.pmd.lang.rule.XPathRule` with `net.sourceforge.pmd.lang.rule.xpath.XPathRule`

**Comment by NickTracker** [comment]
Trying this now

**Comment by NickTracker** [comment]
I'm still seeing the same 

```
stderr:[ERROR] Error at /Users/nzozaya/Salesforce/strk-core/strk-apex-standards.xml:197:9
 195|         language="apex"
 196|         message="Test methods should be preceded by a formal comment explaining their purpose and context."
 197|         class="net.sourceforge.pmd.lang.rule.XPathRule">
              ^^^^^ net.sourceforge.pmd.lang.rule.XPathRule

 198|         <description>
 199|             Ensures that each test method annotated with @isTest has a preceding formal comment.: net.sourceforge.pmd.lang.rule.XPathRule
stderr:[ERROR] Cannot load ruleset /Users/nzozaya/Salesforce/strk-core/strk-apex-standards.xml: 8 XML validation errors occurred
Failed Exit Code: 1
```

My new xml ruleset is - 
```
<?xml version="1.0"?>
<!-- 
    Rules for enforcing our Dev Standards
https://sitetraker.atlassian.net/wiki/spaces/MMEP/pages/2874114053/Development+and+Test+Standards
 -->
<ruleset name="NSYNC Apex Standards"
    xmlns="http://pmd.sourceforge.net/ruleset/2.0.0"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xsi:schemaLocation="http://pmd.sourceforge.net/ruleset/2.0.0 https://pmd.sourceforge.io/ruleset_2_0_0.xsd">
    <description>
        These rules enforce the NSYNC team's documented best practices at compile time.
    </description>
    <rule name="DescriptiveVariableNameRule"
        language="apex"
        message="Use more descriptive variable names rather than short forms."
        class="net.sourceforge.pmd.lang.rule.xpath.XPathRule">
        <description>
            Check for variable name under 4 characters or acronyms following one or less initial small-case letters (https://sitetraker.atlassian.net/wiki/spaces/MMEP/pages/2874114053/Development+and+Test+Standards#1.-Be-descriptive-with-variable-names.)
        </description>
        <priority>2</priority>
        <properties>
            <property name="xpath">
                <value>
                    <![CDATA[
                    //VariableDeclaration
                    [
                        string-length(@Image) < 4
                        or
                        matches(@Image, '^[a-z]{0,1}[A-Z]')
                    ]
                    ]]>
                </value>
            </property>
        </properties>
    </rule>
    <rule name="DetectEarlyReturnOpportunities"
        language="apex"
        message="Consider refactoring this 'if' statement for early return to simplify method structure."
        class="net.sourceforge.pmd.lang.rule.xpath.XPathRule">
        <description>
            This rule detects 'if' statements within methods that are followed immediately by a 'ReturnStatement'. These patterns are candidates for refactoring to use early returns, improving readability and reducing nesting.
        </description>
        <priority>2</priority>
        <properties>
            <property name="xpath">
                <value>
                <![CDATA[
                //Method/BlockStatement/IfElseBlockStatement[
                    following-sibling::*[1][self::ReturnStatement]
                ]
                ]]>
            </value>
            </property>
        </properties>
    </rule>
    <rule name="EnsureProperReturnTypeForApiMethods"
        language="apex"
        message="Methods annotated with @RemoteAction or @AuraEnabled must have a Model or StResult return
    type"
        class="net.sourceforge.pmd.lang.rule.xpath.XPathRule">
        <description>
            This rule enforces a naming convention on the return types of methods annotated with @RemoteAction or @AuraEnabled. The return type should either end with 'Model' or be exactly 'StResult' (https://sitetraker.atlassian.net/wiki/spaces/MMEP/pages/2874114053/Development+and+Test+Standards#4.-Always-append-Model-to-Apex-classes-that-are-going-to-be-used-to-move-data-from-Apex-to-a-frontend%E2%80%A6these-should-be-standalone-classes-to-promote-reusability-of-data-structures-across-multiple-UIs) (https://sitetraker.atlassian.net/wiki/spaces/MMEP/pages/2874114053/Development+and+Test+Standards#5.-Always-prefer-using-Model-Apex-classes-over-String-or-Salesforce-library-structures-like-SObject-or-Schema.*-to-move-data-from-Apex-to-a-frontend...)
        </description>
        <priority>2</priority>
        <properties>
            <property name="xpath">
                <value>
                <![CDATA[
                //Method[
                    (
                        ModifierNode/Annotation[@Name='RemoteAction' or @Name='AuraEnabled']
                    ) and (
                        not(@ReturnType = 'StResult' or contains(@ReturnType, 'Model'))
                    )
                ]
                ]]>
            </value>
            </property>
        </properties>
    </rule>
    <rule name="DetectEmptyListReturnsInAuraAndRemoteActionIfBlocks"
        language="apex"
        message="Within methods annotated with @AuraEnabled or @RemoteAction, prefer throwing exceptions in response to unexpected inputs."
        class="net.sourceforge.pmd.lang.rule.xpath.XPathRule">
        <description>
            This rule flags 'if' blocks within methods annotated with @AuraEnabled or @RemoteAction that return blank lists or null values which may be misleading to invokers of the API (https://sitetraker.atlassian.net/wiki/spaces/MMEP/pages/2874114053/Development+and+Test+Standards#6.-Throw-descriptive-errors-when-you-for-unexpected-inputs-to-let-invokers-of-your-logic-know-that-they-aren%E2%80%99t-invoking-your-method%2FAPI-with-the-correct-parameters.)
        </description>
        <priority>3</priority>
        <properties>
            <property name="xpath">
                <value>
                <![CDATA[
                //Method[
                    ModifierNode/Annotation[@Name='AuraEnabled' or @Name='RemoteAction']
                ]
                /BlockStatement//IfElseBlockStatement/IfBlockStatement[
                    .//ReturnStatement[
                        NewListInitExpression or LiteralExpression[@LiteralType='NULL']
                    ]
                ]
                ]]>
            </value>
            </property>
        </properties>
    </rule>
    <rule name="EnforceSecureDML"
        language="apex"
        message="Direct DML operations should be avoided in favor of Database methods using AccessLevel.USER_MODE or StDmlUtils to ensure security checks are performed."
        class="net.sourceforge.pmd.lang.rule.xpath.XPathRule">
        <description>
        This rule highlights direct usage of DML operations like insert, update, delete, and upsert that are not protected by security-enhancing utility methods like Database actions or StDmlUtils. Highlighting specific DML statements helps developers quickly identify and rectify potentially unsafe code. (https://sitetraker.atlassian.net/wiki/spaces/MMEP/pages/2874114053/Development+and+Test+Standards#2.-Use-AccessLevel.USER_MODE-in-your-DML-calls%2C-it-will-closely-simulate-the-access-constraints-the-real-user-will-face)
    </description>
        <priority>3</priority>
        <properties>
            <property name="xpath">
                <value><![CDATA[
                //Method[
                    not(
                        .//MethodCallExpression[
                            PrimaryExpression/Name[@Name='Database' or contains(@Name, 'StDmlUtils')]
                        ]
                    )
                ]/(
                    .//DmlInsertStatement |
                    .//DmlUpdateStatement |
                    .//DmlDeleteStatement |
                    .//DmlUpsertStatement
                )
            ]]></value>
            </property>
        </properties>
    </rule>
    <!-- Apex test rules -->
    <rule name="RequireTestStartTestInTestMethods"
        language="apex"
        message="Ensure that test methods annotated with @isTest invoke Test.startTest() to properly manage test context and governor limits"
        class="net.sourceforge.pmd.lang.rule.xpath.XPathRule">
        <description>
            This rule ensures that Apex test methods annotated with @isTest include an invocation of Test.startTest(). This practice is crucial for managing governor limits effectively by marking the test's main execution context.
        </description>
        <priority>3</priority>
        <properties>
            <property name="xpath">
                <value>
                <![CDATA[
                //Method[
                    ModifierNode/Annotation[@Name='IsTest']
                    and not(
                        .//MethodCallExpression[
                            @MethodName='startTest'
                        ]
                    )
                ]
                ]]>
            </value>
            </property>
        </properties>
    </rule>
    <rule name="EnsureRunAsInTestMethods"
        language="apex"
        message="Test methods annotated with @isTest should include a System.runAs block to ensure FLS and CRUD is respected within specific user contexts."
        class="net.sourceforge.pmd.lang.rule.xpath.XPathRule">
        <description>
        This rule checks that Apex test methods annotated with @isTest properly simulate user contexts by including System.runAs method calls. This is essential for testing features that depend on user permissions.
    </description>
        <priority>3</priority>
        <properties>
            <property name="xpath">
                <value>
                <![CDATA[
                //Method[
                    ModifierNode/Annotation[@Name='IsTest'] and
                    not(
                        .//RunAsBlockStatement
                    )
                ]
                ]]>
            </value>
            </property>
        </properties>
    </rule>
    <rule name="RequireDocumentationForTestMethods"
        language="apex"
        message="Test methods should be preceded by a formal comment explaining their purpose and context."
        class="net.sourceforge.pmd.lang.rule.xpath.XPathRule">
        <description>
        Ensures that each test method annotated with @isTest has a preceding formal comment. This comment is essential for explaining the test's purpose, providing context, or linking to relevant issue trackers.
    </description>
        <priority>3</priority>
        <properties>
            <property name="xpath">
                <value><![CDATA[
                    //Method[
                        ModifierNode/Annotation[@Name='IsTest'] and
                        not(
                            FormalComment[
                                contains(@Image, 'https') and 
                                contains(@Image, '|')
                            ]
                        )
                    ]             
            ]]></value>
            </property>
        </properties>
    </rule>


</ruleset>
```

I'm happy to take this to the core PMD repo but just want to ensure it belongs there before I do.

**Comment by NickTracker** [comment]
Upon reviewing my error message it seems like the PMD is still citing the old xpath, digging...

**Comment by NickTracker** [comment]
changing the xpath class fixed it, thank you
